File: Javascript/AlphabeticalAddition.js
// Language: Javascript
// Level: 7kyu
// Name of Problem: Alphabetical Addition

// Instructions: Your task is to add up letters to one letter.
// The function will be given a variable amount of arguments, each one being a letter to add.
// Letters will always be lowercase.
// Letters can overflow (see second to last example of the description)
// If no letters are given, the function should return 'z'

// Examples:
// addLetters('a', 'b', 'c') = 'f'
// addLetters('a', 'b') = 'c'
// addLetters('z') = 'z'
// addLetters('z', 'a') = 'a'
// addLetters('y', 'c', 'b') = 'd' // notice the letters overflowing
// addLetters() = 'z'

// Solution 1:
function addLetters(...letters) {
    const alpha = 'zabcdefghijklmnopqrstuvwxy'
    const sum = letters.reduce((sum, letter) => sum + alpha.indexOf(letter), 0) % 26
    return alpha[sum]
  }

// Solution 2:
function addLetters(...letters) {
    if (letters.length === 0) {
      return 'z';
    }
  
    const numbers = letters.map(letter => {
      return letter.charCodeAt() - 96;
    });
    
    const sum = numbers.reduce((a, n) => a + n, 0);
    
    let corrected = sum;
    while (corrected > 26) {
      corrected -= 26;
    }
    
    const newLetter = String.fromCharCode(corrected + 96);
    
    return newLetter;
  }

// Sample Tests Passed:
// const { assert } = require('chai');

// describe("Fixed tests", () => {
//   const tests = [
//     [['a', 'b', 'c'], 'f'],
//     [['z'], 'z'],
//     [['a', 'b'], 'c'],
//     [['c'], 'c'],
//     [['z', 'a'], 'a'],
//     [['y', 'c', 'b'], 'd'],
//     [[], 'z']
//   ];
//   tests.forEach(test => {
//     const str = test[0].map(x => `"${x}"`).join(', ');
//     it(`addLetters(${str})`, () => {
//       assert.strictEqual(addLetters(...test[0]), test[1]);
//     });
//   });
// });

File: C#/AlphabeticalAddition.cs
// Language: C#
// Level: 7kyu
// Name of Problem: Alphabetical Addition

// Instructions: The function will be given an array of letters (chars), each one being a letter to add. 
// Return a char
// Letters will always be lowercase
// Letters can overflow (see second to last example of the description)
// If no letters are given, the function should return 'z'

// Examples: 
// AddLetters(new char[] {'a', 'b', 'c'}) = 'f'
// AddLetters(new char[] {'a', 'b'}) = 'c'
// AddLetters(new char[] {'z'}) = 'z'
// AddLetters(new char[] {'z', 'a'}) = 'a'
// AddLetters(new char[] {'y', 'c', 'b'}) = 'd' // notice the letters overflowing
// AddLetters(new char[] {}) = 'z'

// Solution 1:
//don't forget to uncomment "using System.Linq;" for solution to work

using System;
// using System.Linq;
public class Kata {
  public static char AddLetters(char[] letters) {
  int sum = (letters.Sum(x => x - 96) % 26);
    return sum == 0 ? 'z' : (char)(sum + 96);
  }
}

// Solution 2:
//don't forget to uncomment "using System.Linq;" for solution to work

// using System.Linq;
public class Kata {
public static char AddLetters(char[] letters, string alphabet ="zabcdefghijklmnopqrstuvwxy") => 
  alphabet[letters.Sum(x => alphabet.IndexOf(x)) % 26];
}

// Sample Tests Passed:
// namespace Solution {
//   using NUnit.Framework;
//   using System;
  
//   [TestFixture]
//   public class SolutionTest {    
//     [Test]
//     public void BasicTests() {
//       Assert.AreEqual('f', Kata.AddLetters(new char[] {'a', 'b', 'c'}));
//       Assert.AreEqual('z', Kata.AddLetters(new char[] {'z'}));
//       Assert.AreEqual('c', Kata.AddLetters(new char[] {'a', 'b'}));
//       Assert.AreEqual('c', Kata.AddLetters(new char[] {'c'}));
//       Assert.AreEqual('a', Kata.AddLetters(new char[] {'z', 'a'}));
//       Assert.AreEqual('d', Kata.AddLetters(new char[] {'y', 'c', 'b'}));
//       Assert.AreEqual('z', Kata.AddLetters(new char[] {}));
//     }
//   }
// }